src/core/namespace.c

```diff
@@ -1661,6 +1661,10 @@ static int setup_one_tmp_dir(const char *id, const char *prefix, char **path) {
         if (!x)
                 return -ENOMEM;
 
+        r = mkdir_p(prefix,0777);
```

Comment:
please add a space after the ",", like we usually do.

More importantly though: /var/tmp and /tmp are usually 01777, i.e. have the sticky bit set. You cannot really create those with "mkdir" iirc, hence some care needs to be taken there.

I figure you need to chmod() the dir creating it to set the sticky bit. But only if the dir was actually created by you. If it already exists use the existing mode (because chmod() affects ACLs and stuff and we shouldn't fuck that up if the user set some). But that makes things racy: if two services with PrivateTmp= start simulatenously one might see the dir missing, creates it, now the second is started and sees it existing, and only then the first one gets to change the access mode. Then the second one would run with incorrect access mode which it probably shouldn't.

Hence I think creating the prefix under a random name, then adjusting the mode and moving it should be the right approach. Maybe something like this:

```c
static int make_tmp_prefix(const char *prefix) {
        _cleanup_free_ char *t = NULL;
        int r;

        /* Don't do anything unless we know the dir is actually missing */
        r = access(prefix, F_OK);
        if (r >= 0)
                return 0;
        if (errno != ENOENT)
                return -errno;

        r = mkdir_parents(prefix, 0755);
        if (r < 0)
                return r;

        r = tempfn_random(prefix, NULL, &t);
        if (r < 0)
                return r;

        if (mkdir(t, 0777) < 0)
                return -errno;

       if (chmod(t, 01777) < 0) {
                 r = -errno;
                 (void) rmdir(t);
                 return r;
        }

        if (rename(t, prefix) < 0) {
                r = -errno;
                (void) rmdir(t);
                return r == -EEXIST ? 0 : r; /* it's fine if someone else created the dir by now */
        }
                        
        return 0; 

}

Reply:
wow, I'm always amazed by how complex that sort of things are and by the people that  actually understand it.

I'll give your code a whirl and resubmit...